Protocols: Add unit test cases for messages (#298)
pez-globo/pufferfish-vent-software: firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Protocols/messages.txt

/* Write function

> Scenario: Message: The write method correctly generates an encoded body from the protobuf message descriptor payload

Reviewer: All tests in this scenario need to describe what the value of the `type` and `payload` members should be after `write` is called.

>   GIVEN("A Message object constructed with StateSegment Taggedunion, default message types and a max size of 252 bytes")

Reviewer: Since no explicit constructor is defined for `Message`, I think it's more accurate to say "A Message object default-constructed with ... as its class template parameters". Or you could use this GIVEN to also initialize the values of the `type` and `payload` members.

  // assertions carried out in given:
  // output buffer is empty

    WHEN("The value of the message payload tag is greater than the size of the message descriptor array")
      THEN("The write method reports invalid type status")
      THEN("After the write method is called, the output buffer remains unchanged")

    WHEN("The message payload tag is such that it corresponds to Unrecognized message descriptor in descriptor array")
      THEN("The write method reports invalid type status")
      THEN("After the write method is called, the output buffer remains unchanged")

>     WHEN("The output buffer is too small to hold the generated body")

Reviewer: Either:
1. In this WHEN and the next WHEN, you should describe the values of the `type` and `payload` members, or
2. You could split these WHENs into another GIVEN which describes the values of those members.

      THEN("The write method reports invalid length status")
      THEN("After the write method is called, the output buffer remains unchanged")

    WHEN("The size of the output buffer is zero")
      THEN("The write method reports invalid length status")

>     // Intent: To show that the write method does not change the internal "type" member of the message class

Reviewer: I'm surprised, because the first line of the `write` method is: `auto type = static_cast<uint8_t>(payload.tag);`. I must be missing/misunderstanding something here

Reviewer: Let's make the `Message::write` method's interface more consistent with the `Datagram::write` and `CRCElement::write` method interfaces. So the contact the method should uphold is that it will update the type member, and the method is not const. This also means that `MessageSender::transform` should take a `const TaggedUnion &payload` (where `TaggedUnion` is now a template parameter for `MessageSender`) instead of `const Message &input_message`; and `BackendSender::transform` should similary take a `const Application::StateSegment &state_segment` instead of a `const BackendMessage &input_message`. Then `Backend::output`, instead of instantiating `BackendMessage message` and calling `synchronizer_.output(message.payload)`, should instantiate an `Application::StateSegment state_segment` and call `synchronizer_.output(state_segment)`.

Reviewer: `MessageSender::transform` will need to instantiate its own `Message input_message` object, and then set `input_message.payload = input_payload`, and hopefully we get copy-by-value (we should verify this by checking that the value of `input_payload` hasn't changed after calling `transform`).

    WHEN("The internal type member is set to an arbitrary value and then write method is called")
    // setup:
    // type is set to 4
    // A sensor measurements payload message is set, which has value of type as 2
      THEN("The write method reports ok status")
      THEN("After the write method is called, type member remains unchanged)

>   GIVEN("A Message object constructed with StateSegment Taggedunion, default message types and a max size of 126 bytes")

Reviewer: It looks like I might not be using the `max_size` template parameter anywhere in the implementation of the `Messages` classes. I would suggest adding `static_asserts` comparing `max_size` against the sizes of the `input_size` and/or `output_size` method template parameters into the definitions of the `parse` and `write` methods.

    WHEN("The payload is a sensor measurements message and write method is called")
      THEN("The write method reports ok status")
      THEN("The body's header correctly stores the value of message type obtained from the message payload tag")

>       THEN("The output buffer is as expected")

Reviewer: I mentioned in the review of #297 that it could be good to describe the output buffer (e.g. by giving its expected length), and I think it would be useful here too

    WHEN("The sensor measurments message data is written")
      THEN("The write method reports ok status")
      THEN("The body's header correctly stores the value of message type obtained from the message payload tag")
      THEN("The output buffer is as expected")

/* Write function for each of the message types (example: sensor_measurements)
    WHEN("The payload is a sensor measurements message and write method is called")
    // setup: sensor_measurements.flow = 50
      THEN("The write method reports ok status")
      THEN("The body's header correctly stores the value of message type obtained from the message payload tag")
      THEN("The output buffer is as expected")
/*

/* Parse function

Scenario: Message: The parse method correctly updates internal message type and sets the message payload fields after decoding from the input buffer

  GIVEN("A Message object constructed with StateSegment Taggedunion, default message types and a max size of 252 bytes")

    WHEN("A body with an encoded payload and 1 byte header of value equal to 0 is parsed")
      THEN("The parse method reports invalid type status")

>       THEN("After the parse method is called, The value assigned to the type member is equal to the input_buffer body's header")

Reviewer: The header will be changed to add a reserved field for future use, so this THEN should specify that it's comparing against the type field of the header.

      THEN("The payload Taggedunion tag is set to Unknown")
      THEN("The message payload fields are not set")

    WHEN("An empty input buffer body is parsed")
      THEN("The parse method reports invalid length status")

>       THEN("The payload Taggedunion tag is set to Unknown")

Reviewer: I don't think `parse` sets the internal members if it returns `invalid_length`, I think it just leaves them at whatever their previous values were. If this is accurate, we should design tests to check that it actually does leave them at the previous values (this will involve adding a new GIVEN with different initial values of the members, and maybe the GIVENs should be reorganized under a separate scenario)

Reviewer: Did you find that parse sets the internal members, such as `tag`, even if the method returns `invalid_length`?

Author: No, it does not, it remains unchanged as the payload is uninitialized in this GIVEN, it is unknown before or after parse

      THEN("The message payload fields are not set")

    WHEN("A body with an empty payload and 1 byte header of value equal to the message descriptor array size is parsed")
      THEN("The parse method reports invalid type status")
      THEN("After the parse method is called, The value assigned to the type member is equal to the input_buffer body's header")

>       THEN("The payload Taggedunion tag is set to Unknown")

Reviewer: There are three ways `parse` can return `invalid_type`; in one of them, it does not change `payload.tag`, and in the other two, it depends on the GIVEN/WHEN (this will vary depending on the `MessageTypes` template parameter and the `pb_protobuf_descriptors` argument). We should cover each of the three ways `invalid_type` can be returned. It could make sense for this to be a separate scenario, because it may involve multiple GIVENs.

Reviewer: Where is the test of `invalid_type` resulting from `!MessageTypes::includes(type)`?

Author: So the tests where the body with `header` value greater than and equal to descriptor array size (8), both actually failed this test instead, as `MessageTypes` enum has enums till 7. Therefore I have re-structured these tests now

      THEN("The message payload fields are not set")

    WHEN("A body with an empty payload and 1 byte header of value greater than the message descriptor array size is parsed")
      THEN("The parse method reports invalid type status")
      THEN("After the parse method is called, The value assigned to the type member is equal to the input_buffer body's header")
      THEN("The payload Taggedunion tag is set to Unknown")
      THEN("The message payload fields are not set")

    WHEN("A body with a 1 byte header and payload of sensor_measurements message with invalid encoding is parsed")
      THEN("The parse method reports invalid encoding status")
      THEN("After the parse method is called, The value assigned to the type member is equal to the input_buffer body's header")
      THEN("The payload Taggedunion tag is set to sensor_measurements")

>       THEN("The message payload fields are not set")

Reviewer: I think this contradicts the previous line, as `payload.tag` is a field of `payload`. Maybe you mean that the `payload.value` field is not set?

>     WHEN("The input buffer size is greater than the payload max size")

Reviewer: This doesn't make sense to me, because Messages.tpp never refers to `payload_max_size`.

      THEN("The status of the parse function returns invalid encoding")

/* test cases for parse functions for each message type

>     WHEN("A body with 1 byte header and a paylod of {Message type} message type is parsed")

Reviewer: I think this underspecifies the contents of the header - the type field should be consistent with the payload.

Reviewer: To rephrase: What's the value of the 1-byte header in each WHEN?

      THEN("The parse method reports ok status")
      THEN("After the parse method is called, The value assigned to the type member is equal to the input_buffer body's header")
      THEN("The payload Taggedunion tag is set to {Message type}")

>       THEN("The message payload field values are as expected")

Reviewer: The language is a bit ambiguous. I think rewording it will resolve this: "THEN the payload.values field has the expected {Message type} field values" or "THEN the payload field's values field has the expected {Message type} field values" or something similar.

Whatever you choose, let's use the same pattern when we refer to the payload.tag field (alternately, "the payload field's tag field" or something similar).

Reviewer: It looks like this comment was marked as resolved without addressing either of the suggestions in this comment?

Author: I had updated everywhere else and forgot to update here. fixed now.

/* Write function
Scenario: Message: The write method correctly generates an encoded body from the protobuf message descriptor payload

  GIVEN("A Message object default-constructed with StateSegment Taggedunion, default message types and a max size of 252 bytes as its class template parameters")
  // assertions carried out in given:
  // message class type member is 0

    // setup:
    // message class type member set to 5
    // payload of parameters_request is set
    // message descriptor array containing one element corresponding to Unknown message

    WHEN("Write method is called on a message object with it's value of the payload tag greater than the size of the descriptor array")
      THEN("The write method reports invalid type status")

>       THEN("After the write method is called, the type member remains unchanged")

Reviewer: Now that we removed `auto type = ...`, the type member should be changed on every `write` method call, right? (This applies to all tests of the write method)

>       THEN("The payload.values field data remains unchanged")

Reviewer: Does the `payload.tag` field remain unchanged too? (This applies to all tests of the write method)

>       THEN("The output buffer remains empty")

Reviewer: I think this should be "the output buffer remains unchanged"?

Author: fixed.

    // setup:
    // message class type member is unchanged
    // payload tag is set to Unknown

>     WHEN("Write method is called on a message object where it's payload tag value is such that it corresponds to Unrecognized message descriptor in descriptor array")

Reviewer: ```suggestion
    WHEN("Write method is called on a message object whose payload.tag value corresponds to an Unrecognized message descriptor in the descriptor array")
```

      THEN("The write method reports invalid type status")
      THEN("After the write method is called, the type member remains unchanged")
      THEN("The payload.values field data remains unchanged")
      THEN("The output buffer remains empty")

  GIVEN("A Message object default-constructed with StateSegment Taggedunion, default message types and a max size of 126 bytes as its class template parameters")
    // setup:
    // sensor_measurements.flow = 30;
    // sensor_measurements.fio2 = 85;
    // sensor_measurements.spo2 = 72;  
    WHEN("The payload is a sensor measurements message and write method is called")
      THEN("The write method reports ok status")

>       THEN("After the write method is called, the type member is set to the value of the message payload tag")

Reviewer: Does the `payload.value` field remain unchanged too? (This applies to all tests of the write method)

Author: fixed.

      THEN("The type field of the body's header correctly stores the value of message type obtained from the message payload tag")
      THEN("The output buffer has an expected sequence of 16 bytes")

  GIVEN("A Message object constructed with StateSegment Taggedunion, default message types and a max size of 508 bytes")
    // setup:
    // parameters.fio2 = 60;
    // parameters.mode = VentilationMode_hfnc;
    // parameters.ventilating = true;
    WHEN("The parameters message data is written")
      THEN("The write method reports ok status")
      THEN("After the write method is called, the type member is set to the value of the message payload tag")
      THEN("The type field of the body's header correctly stores the value of message type obtained from the message payload tag")
      THEN("The output buffer has an expected sequence of 10 bytes")

> 

Reviewer: We're still missing a test for the `write` method returning `invalid_length`, which is one of the three possible error codes returned by this method. I believe there are two cases for this: 1) less than header size, and 2) greater than header size but less than the required size (e.g. for the sensor_measurements test which should generate a 16-byte output, setting max_size to 14 bytes).

If we're unable to generate tests which trigger `invalid_encoding`, we should add a comment to that effect in the test files, and explaining what we tried and why we think we won't be able to generate such a test.

Author: > We're still missing a test for the `write` method returning `invalid_length`, which is one of the three possible error codes returned by this method. I believe there are two cases for this: 1) less than header size, and 2) greater than header size but less than the required size (e.g. for the sensor_measurements test which should generate a 16-byte output, setting max_size to 14 bytes).

added.

> If we're unable to generate tests which trigger `invalid_encoding`, we should add a comment to that effect in the test files, and explaining what we tried and why we think we won't be able to generate such a test.

will do.

    // setup:
    // sensor_measurements.flow = 30;
    // sensor_measurements.fio2 = 85;
    // sensor_measurements.spo2 = 72;
    WHEN("The payload is a sensor measurements message and write method is called")
      THEN("The write method reports ok status")
      THEN("After the write method is called, the type member is equal to the value of sensor_measurements tag")
      THEN("The type field of the body's header correctly stores the value of message type obtained from the message payload tag")
      THEN("The output buffer has an expected sequence of 16 bytes")
/*

/* Parse function

Scenario: Message: The parse method correctly updates internal message type and sets the message payload fields after decoding from the input buffer

  GIVEN("A Message object constructed with StateSegment Taggedunion, default message types and a max size of 252 bytes")
    // setup:
    // message class type member is unchanged
    // payload is uninitialised

    WHEN("An empty input buffer body is parsed")
      THEN("The parse method reports invalid length status")
      THEN("The payload Taggedunion tag is set to Unknown")

>       THEN("the payload.values field remains unchanged")

Reviewer: What happens to the `type` field after the method call? (This applies to all parse and MessageReceiver tests)

Author: fixed.

>     WHEN("A body with an encoded payload and 1 byte header of value equal to 0 is parsed")

Reviewer: It'd be helpful if you put the WHENs for the tests of the other ways we can get `invalid_type` return code above this WHEN, since this WHEN is constructed so that `MessageTypes::includes(type) == true && type < pb_protobuf_descriptors.size()`; this way, the order of tests matches the order of code in the implementation.

Author: I think the order should be fine now.

      THEN("The parse method reports invalid type status")
      THEN("After the parse method is called, The value assigned to the type member is equal to the type field of the input_buffer body's header")
      THEN("The payload Taggedunion tag is set to Unknown")
      THEN("the payload.values field remains unchanged")

    WHEN("A body with an empty payload and 1 byte header of value equal to the message descriptor array size is parsed")
      THEN("The parse method reports invalid type status")
      THEN("After the parse method is called, The value assigned to the type member is equal to the type field of the input_buffer body's header")

>       THEN("The payload Taggedunion tag is set to Unknown")

Reviewer: In this test of `invalid_type` resulting from `type >= pb_protobuf_descriptors.size()`, shouldn't `tag` be set to the value `pb_protobuf_descriptors.size() + 1` by the method?

Author: yes. fixed.

      THEN("the payload.values field remains unchanged")

    WHEN("A body with an empty payload and 1 byte header of value greater than the message descriptor array size is parsed")
      THEN("The parse method reports invalid type status")
      THEN("After the parse method is called, The value assigned to the type member is equal to the type field of the input_buffer body's header")

>       THEN("The payload Taggedunion tag is set to Unknown")

Reviewer: In this test of `invalid_type` resulting from `type >= pb_protobuf_descriptors.size()`, shouldn't `tag` be set to the value `pb_protobuf_descriptors.size() + 1` by the method?

Author: yes. fixed.

      THEN("the payload.values field remains unchanged")

>     WHEN("A body with a 1 byte header and payload of sensor_measurements message with invalid encoding is parsed")

Reviewer: What's the value of the 1-byte header?

      THEN("The parse method reports invalid encoding status")
      THEN("After the parse method is called, The value assigned to the type member is equal to the input_buffer body's header")
      THEN("The payload Taggedunion tag is set to sensor_measurements")
      THEN("The message payload fields are not set")

/* test cases for parse functions for each message type
    WHEN("A body with 1 byte header and a paylod of {Message type} message type is parsed")
      THEN("The parse method reports ok status")
      THEN("After the parse method is called, The value assigned to the type member is equal to the input_buffer body's header")
      THEN("The payload Taggedunion tag is set to {Message type}")
      THEN("The message payload field values are as expected")
      THEN("The input buffer is unchanged after parse")
/*

> Scenario: The parse method correctly parses bodies when the payload and field members have data

Reviewer: Right now the test under this isn't about correctly parsing bodies, but about error-handling when parsing fails. However, we should also have a test about correctly parsing bodies when the payload and field members have data.

>   GIVEN("A Message object constructed with StateSegment Taggedunion, default message types and a max size of 252 bytes")

Reviewer: Was there a reason you left `payload.tag` uninitialized?

Author: Since I use the `set` method to initialize the payload, I forgot to mention explicitly that the tag is set to something. I fixed it.

  // setup:
  // type = 5
  // parameters_request.fio2 = 80;
  // parameters_request.mode = VentilationMode_hfnc;
  // parameters_request.ventilating = true;

    WHEN("An empty input buffer body is parsed")
      THEN("The parse method reports invalid length status")
      THEN("After the parse method is called, The value assigned to the type member remains unchanged")

>       THEN("The payload Taggedunion tag is set to Unknown")

Reviewer: Are you sure that the value of the tag is changed to be unknown, instead of being left unchanged?

Author: Ya my bad, it is left unchanged.

    // setup:
    // message class type member is unchanged
    // payload is uninitialised

    WHEN("An empty input buffer body is parsed")
      THEN("The parse method reports invalid length status")
      THEN("After the parse method is called, the type member is set to 0")
      THEN("The payload.tag field remains unchanged")
      THEN("the payload.values field remains unchanged")

    WHEN("A body with an empty payload and 1-byte header whose value is not included in MessageTypes enum")
      THEN("The parse method reports invalid type status")
      THEN("After the parse method is called, The value assigned to the type member is equal to the type field of the input_buffer body's header")
      THEN("The payload.tag field remains unchanged")
      THEN("the payload.values field remains unchanged")

    WHEN("A body with an empty payload and 1-byte header of value equal to the message descriptor array size is parsed")
    // setup:
    // custom descriptor array of size 4
      THEN("The parse method reports invalid type status")
      THEN("After the parse method is called, The value assigned to the type member is equal to the type field of the input_buffer body's header")

>       THEN("The payload.tag field is set to parameters")

Reviewer: I don't understand the relationship between payload.tag == parameters and the size of the message descriptor array. Maybe there's some information in the test code which isn't described in the test description? (similar for the next WHEN with payload.tag == parameters_request)

Author: Right the information in the test code would be the header of the input_buffer which in this case is `0x04` (which is equal to the descriptor array size), the reason I created a different descriptor array rather than using the one from the backend for these two cases was that to return `invalid_type` on an array of 8 elements, `MessageTypes::includes(type)` would return `invalid_type` instead.  therefore created an array with 4 elements, and added the test cases for equal to and greater than the array size.

And also that's why the `payload.tag is set to parameters`  same is the case for the next WHEN, where the header is`0x05`. so I'm just asserting that the payload.tag parameter is updated.

Reviewer: I see, so the context here is "THEN(The payload.tag field is set to parameters, _which corresponds to value 0x04_)"
